peer.go

```diff
@@ -4,134 +4,260 @@ import (
 	"fmt"
 	"io"
 	"net"
+	"time"
 
-	. "github.com/tendermint/go-common"
+	cmn "github.com/tendermint/go-common"
 	cfg "github.com/tendermint/go-config"
-	"github.com/tendermint/go-wire"
+	crypto "github.com/tendermint/go-crypto"
+	wire "github.com/tendermint/go-wire"
 )
 
+// Peer could be marked as persistent, in which case you can use
+// Redial function to reconnect. Note that inbound peers can't be
+// made persistent. They should be made persistent on the other end.
+//
+// Before using a peer, you will need to perform a handshake on connection.
 type Peer struct {
-	BaseService
+	cmn.BaseService
 
 	outbound bool
-	mconn    *MConnection
+
+	conn  net.Conn     // source connection
+	mconn *MConnection // multiplex connection
+
+	authEnc    bool // authenticated encryption
+	persistent bool
+	config     cfg.Config
 
 	*NodeInfo
 	Key  string
-	Data *CMap // User data.
+	Data *cmn.CMap // User data.
+}
+
+func newPeer(addr *NetAddress, reactorsByCh map[byte]Reactor, chDescs []*ChannelDescriptor, onPeerError func(*Peer, interface{}), config cfg.Config, privKey crypto.PrivKeyEd25519) (*Peer, error) {
+	conn, err := dial(addr, config)
+	if err != nil {
+		return nil, err
+	}
+
+	// outbound = true
+	return newPeerFromExistingConn(conn, true, reactorsByCh, chDescs, onPeerError, config, privKey)
+}
+
+func newPeerFromExistingConn(conn net.Conn, outbound bool, reactorsByCh map[byte]Reactor, chDescs []*ChannelDescriptor, onPeerError func(*Peer, interface{}), config cfg.Config, privKey crypto.PrivKeyEd25519) (*Peer, error) {
+	// Encrypt connection
+	if config.GetBool(configKeyAuthEnc) {
+		var err error
+		conn, err = MakeSecretConnection(conn, privKey)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	p := &Peer{
+		outbound: outbound,
+		authEnc:  config.GetBool(configKeyAuthEnc),
+		conn:     conn,
+		config:   config,
+		Data:     cmn.NewCMap(),
+	}
+
+	p.mconn = createMConnection(conn, p, reactorsByCh, chDescs, onPeerError, config)
+
+	p.BaseService = *cmn.NewBaseService(log, "Peer", p)
+
+	return p, nil
 }
 
+// CloseConn should be used when the peer was created, but never started.
+func (p *Peer) CloseConn() {
+	p.conn.Close()
+}
+
+// MakePersistent marks the peer as persistent.
+func (p *Peer) MakePersistent() {
+	if !p.outbound {
+		panic("inbound peers can't be made persistent")
+	}
+
+	p.persistent = true
+}
+
+// IsPersistent returns true if the peer is persitent, false otherwise.
+func (p *Peer) IsPersistent() bool {
+	return p.persistent
+}
+
+// HandshakeTimeout performs a handshake between a given node and the peer.
 // NOTE: blocking
-// Before creating a peer with newPeer(), perform a handshake on connection.
-func peerHandshake(conn net.Conn, ourNodeInfo *NodeInfo) (*NodeInfo, error) {
+func (p *Peer) HandshakeTimeout(ourNodeInfo *NodeInfo, timeout time.Duration) error {
+	// Set deadline for handshake so we don't block forever on conn.ReadFull
```

Comment:
Ok ... now I'm not sure we need this here. Would ReadBinary block forever or just fail? Whether or not we need it here, we definetely need one before we authenticate/encrypt the connection. I can add that